core/src/main/java/org/elasticsearch/action/admin/indices/shards/IndicesShardsStoresResponse.java

```diff
@@ -0,0 +1,383 @@
+            }
+        }
+
+        private StoreStatus() {
+        }
+
+        public StoreStatus(DiscoveryNode node, long version, Allocation allocation, Throwable storeException) {
+            this.node = node;
+            this.version = version;
+            this.allocation = allocation;
+            this.storeException = storeException;
+        }
+
+        /**
+         * Node the shard belongs to
+         */
+        public DiscoveryNode getNode() {
+            return node;
+        }
+
+        /**
+         * Version of the shard, used to select the store that will be
+         * used as a primary.
+         */
+        public long getVersion() {
+            return version;
+        }
+
+        /**
+         * Exception while trying to open the
+         * shard index or from when the shard failed
+         */
+        public Throwable getStoreException() {
+            return storeException;
+        }
+
+        /**
+         * The allocation status of the store.
+         * {@link Allocation#PRIMARY} indicates a primary shard copy
+         * {@link Allocation#REPLICA} indicates a replica shard copy
+         * {@link Allocation#UNUSED} indicates an unused shard copy
+         */
+        public Allocation getAllocation() {
+            return allocation;
+        }
+
+        static StoreStatus readStoreStatus(StreamInput in) throws IOException {
+            StoreStatus storeStatus = new StoreStatus();
+            storeStatus.readFrom(in);
+            return storeStatus;
+        }
+
+        @Override
+        public void readFrom(StreamInput in) throws IOException {
+            node = DiscoveryNode.readNode(in);
+            version = in.readLong();
+            allocation = Allocation.readFrom(in);
+            if (in.readBoolean()) {
+                storeException = in.readThrowable();
+            }
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            node.writeTo(out);
+            out.writeLong(version);
+            allocation.writeTo(out);
+            if (storeException != null) {
+                out.writeBoolean(true);
+                out.writeThrowable(storeException);
+            } else {
+                out.writeBoolean(false);
+            }
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            node.toXContent(builder, params);
+            builder.field(Fields.VERSION, version);
+            builder.field(Fields.ALLOCATED, allocation.value());
+            if (storeException != null) {
+                builder.startObject(Fields.STORE_EXCEPTION);
+                ElasticsearchException.toXContent(builder, params, storeException);
+                builder.endObject();
+            }
+            return builder;
+        }
+
+        @Override
+        public int compareTo(StoreStatus other) {
+            if (other == null) {
```

Comment:
Out of curiosity is null a valid incoming value here? Is it required by the compareTo semantics and if so, does null need to be the highest value?